Added support to create and delete multiple databases in MySQL

pratikgadiya12 opened the pull request:
##### SUMMARY
Creation/Deletion of multiple databases now supported in MySQL

Fixes: #58370

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/database/mysql/mysql_db.py

pratikgadiya12 committed:
Added support to create/delete mulitiple databases in MySQL

Fixes: #58370

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -78,6 +78,18 @@
 
+- name: Create new database with name 'foo', 'bar'
+  mysql_db:
+    name: foo,bar
```
I would only add the second example. They differ only in a syntax detail which is not that important.

pratikgadiya12 replied:
Ack, fixing it

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -78,6 +78,18 @@
+    name: foo,bar
+    state: present
+
+- name: Create new database with name 'foo', 'bar'
```
```suggestion
- name: Create new databases with names 'foo' and 'bar'
```

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -115,6 +127,23 @@
+    name: foo,bar
+    state: absent
+
+- name: Delete database with name 'foo', 'bar'
```
```suggestion
- name: Make sure there is neither a databases with name 'foo', nor one with name 'bar'
```

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -136,13 +165,16 @@
+    if not db:
+        return False
+    for each_db in db:
+        query = "DROP DATABASE %s" % mysql_quote_identifier(each_db, 'database')
```
What happens if the database does not exist?

pratikgadiya12 replied:
Here there are 2 cases,

1. If non-existing `db` is **passed** via ansible playbook then it silently passes since `existence_list` variable would be `[]` returning `False`

```
$ ansible-playbook delete_nonexisting_db.yml 
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [Sample] *****************************************************************************************************************************************************************************************************

TASK [Delete databases db1 and db3] *******************************************************************************************************************************************************************************
ok: [localhost]

TASK [assert successful completion of deleting existing database] *************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

TASK [run command to list databases like specified database name] *************************************************************************************************************************************************
changed: [localhost]

TASK [assert that specific databases which were deleted does not exist] *******************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

2. If non-existing `db` is **not passed** via ansible playbook then it will throw an error
`FAILED! => {"changed": false, "msg": "error deleting database: (1008, u\"Can't drop database 'abcde'; database doesn't exist\")"}`

How do we want it to behave?

felixfontein replied:
What do you mean by "not passed"? If the DB is not specified as a module option, why does the code tries to delete it?

pratikgadiya12 replied:
I was thinking on the lines where this function could be re-used by some other scripts (and not ansible or its modules). In that situation, the function will throw an error saying database does not exist to delete. I am unnecessarily thinking too much here and making it complex.

Short answer to the question "`What happens if the database does not exist?`" is that nothing is changed on the system and ansible playbook executes successfully

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +383,74 @@ def main():
+    if state == "absent":
+        if module.check_mode and not db_exists(cursor, db):
+            module.exit_json(changed=False, db=db_name, db_list=db)
+        elif module.check_mode and db_exists(cursor, db):
```
In case check mode is enabled, you are making the same call to `db_exists` **twice** (in case the first one returns `False`). That's not exactly efficient.

How about something like:
```
if module.check_mode:
    module.exit_json(changed=db_exists(cursor, db), db=db_name, db_list=db)
```

felixfontein replied:
Also, why do you need to call `db_exists` again at all? You already have `existence_list` and `existence_list`, that should be enough for the rest of the module.

felixfontein replied:
Finally, `db_exists` checks whether **all** DBs in the list exist. This doesn't really help here. You really need to work with `existence_list` and `existence_list`.

pratikgadiya12 replied:
Fixing it, thanks

felixfontein commented on test/integration/targets/mysql_db/tasks/main.yml:
```diff
@@ -243,3 +243,5 @@
 
 - include: state_dump_import.yml format_type=bz2 file=dbdata.bz2 format_msg_type=bzip2 file2=dump2.bz2 file3=dump3.bz2
+
+- include: multi_db_create_delete.yml 
```
Do you have tests where some of the DBs you specify exists and some not? If not, you should really add such tests.

pratikgadiya12 replied:
Fixed

felixfontein reviewed (commented):
Please rework the logic using the lists. I'll take another look then.

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -294,6 +330,7 @@ def main():
     db = module.params["name"]
     if not db:
         module.fail_json(msg="Please provide at least one database name")
```
Since a related thing came up in the Ansible project meeting: modules should not fail if nothing is passed. (At least where it makes sense, and here it does: if no DBs are specified, there's simply no change to do.)

Andersson007 commented on test/integration/targets/mysql_db/tasks/multi_db_create_delete.yml:
```diff
@@ -0,0 +1,211 @@
+# Copyright (c) 2018, Pratik Gadiya <pratikgadiya1@gmail.com>
```
```suggestion
# Copyright (c) 2019, Pratik Gadiya <pratikgadiya1@gmail.com>
```

Andersson007 commented on test/integration/targets/mysql_db/tasks/multi_db_create_delete.yml:
```diff
@@ -0,0 +1,211 @@
+- name: assert successful completion of create database
+  assert:
+    that:
+       - "result.changed == true"
```
```suggestion
       - result.changed == true
```
I've heard from a coreteam member that it is a good practice not to use quotes where not necessary ;) but if you like to use it, I don't insist :)

pratikgadiya12 replied:
Fixing it

Andersson007 reviewed (changes requested):
@pratikgadiya12 , thank you for this PR!
Let's start from covering these changes by integration tests. Then I make code and doc reviews.

1. First of all, regarding that @felixfontein said , please, remove tests in a comma-separated form or a list form (on your choice) from the test file because they are interchangeable
2. We should check:
- Create multiple databases check_mode (done)
- Create multiple databases (partially done) - you check it by: ```"'database%' not in mysql_result.stdout"``` it may return true when one db exists, but **we need to be sure that the both databases exist**, so, better to check them by names
- Delete multiple databases check_mode (done)
- Delete multiple databases (done)
- Recreate multiple databases (done)

Please add checks for the following case:
- As I wrote above, when we check that the databases exist on the server instance we need to be sure that each of them exists, change it please (preferably by check them by names).
- Delete one of the databases, result.changed == true
- Check that it doesn't exist and the second db exists
- Try to create both in check_mode, result.changed == true
- Check that the deleted db actually doesn't exist, the second exist
- Try to create both again in actual mode, result.changed == true
- Check that the both exist
- Remove both databases to clean up

pratikgadiya12 committed:
Added additional tests cases and fixed documentation changes

pratikgadiya12 committed:
Code refactoring and added tests for better test coverage

- Removed db_exists usage from most of the code. Used existence_list
 and non_existence_list instead

- Added additional tests to cover all scenarios w.r.t creation and deletion
 on multiple databases

- Added tests for dump operations

pratikgadiya12 committed:
Minor fix

pratikgadiya12 committed:
Minor fix - create check mode test

pratikgadiya12 committed:
Added dump tests for better dump tests coverage

pratikgadiya12 committed:
Removed minor database connection details

pratikgadiya12 committed:
fixed error

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +373,65 @@ def main():
+        if module.check_mode:
+            module.exit_json(changed=not bool(non_existence_list), db=db_name, db_list=db)
+        if non_existence_list:
+            module.fail_json(msg="Cannot dump database %r - not found" % (' '.join(non_existence_list)))
```
```suggestion
            module.fail_json(msg="Cannot dump database(s) %r - not found" % (', '.join(non_existence_list)))
```

pratikgadiya12 replied:
fixed

felixfontein reviewed (commented):
Just looked at the code, looks OK now I think.

pratikgadiya12 committed:
Added test case for import operations

pratikgadiya12 commented:
> @pratikgadiya12 , thank you for this PR!
> Let's start from covering these changes by integration tests. Then I make code and doc reviews.
> 
> 1. First of all, regarding that @felixfontein said , please, remove tests in a comma-separated form or a list form (on your choice) from the test file because they are interchangeable
> 2. We should check:
> 
> * Create multiple databases check_mode (done)
> * Create multiple databases (partially done) - you check it by: `"'database%' not in mysql_result.stdout"` it may return true when one db exists, but **we need to be sure that the both databases exist**, so, better to check them by names
> * Delete multiple databases check_mode (done)
> * Delete multiple databases (done)
> * Recreate multiple databases (done)
> 
> Please add checks for the following case:
> 
> * As I wrote above, when we check that the databases exist on the server instance we need to be sure that each of them exists, change it please (preferably by check them by names).
> * Delete one of the databases, result.changed == true
> * Check that it doesn't exist and the second db exists
> * Try to create both in check_mode, result.changed == true
> * Check that the deleted db actually doesn't exist, the second exist
> * Try to create both again in actual mode, result.changed == true
> * Check that the both exist
> * Remove both databases to clean up

Thanks for the detailed checks @Andersson007 :)

I have incorporated all the checks mentioned above and also added 2 more checks to create and delete operations.

1. After creating databases in actual mode, try to recreate existing databases in check mode. This will help us to test that `changed` value equals to `False` for `check_mode` when `state=present`
2. Similarly, after deleting databases in actual mode, try to again delete already deleted databases in check mode. This will help us to test that `changed` value equals to `False` for `check_mode` when `state=absent`

In addition to this, I have also added multiple tests for dump and import operations to enhance more test coverage w.r.t check_mode and actual mode operations.

pratikgadiya12 commented:
Thanks, @felixfontein, and @Andersson007 for the review.
PR open for review again :)

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -24,7 +24,7 @@
       - I(name=all) May only be provided if I(state) is C(dump) or C(import).
-      - List of databases is provided with I(state=dump) only.
+      - List of databases is provided with I(state=dump), I(state=present) and I(state=absent).
       - if name=all Works like --all-databases option for mysqldump (Added in 2.0).
```
Is this correctly handled? I think right now, you're trying to determine whether a database called "all" exists or not.

pratikgadiya12 replied:
Yes, this has been handled https://github.com/pratikgadiya12/ansible/blob/issue_58370/lib/ansible/modules/database/mysql/mysql_db.py#L355

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -293,7 +319,8 @@ def main():
     if not db:
-        module.fail_json(msg="Please provide at least one database name")
+        module.exit_json(changed=False, db=db)
+    db = [each_db.strip() for each_db in db]
```
It might make sense here to check for `db == ['all']`, at least for `dump` and `import`; if that's true, set `db == []` and `db_all = True` (or something like that).

pratikgadiya12 replied:
This has been handled in https://github.com/pratikgadiya12/ansible/blob/issue_58370/lib/ansible/modules/database/mysql/mysql_db.py#L355

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -318,8 +345,8 @@ def main():
 
-    if len(db) > 1 and state != 'dump':
-        module.fail_json(msg="Multiple databases is only supported with state=dump")
+    if len(db) > 1 and state == 'import':
```
This check would have to be adjusted for `db_all` if you make the above change.

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +373,65 @@ def main():
+                                 exception=traceback.format_exc())
+        module.exit_json(changed=changed, db=db_name, db_list=db)
+    elif state == "dump":
+        if module.check_mode:
```
The next `if` should come first. The module should die if not all DBs exists no matter whether it's in check mode or not.

pratikgadiya12 replied:
I kept the logic same as [previous](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/mysql/mysql_db.py#L420) where it checks first the `check_mode` for dump operation if db does not exist.

If we alter this, it might break some ansible user's playbook

felixfontein replied:
That's true. The old behavior was wrong IMO though, and I would consider it a bug. Check mode should be as close as possible to the real behavior of the module, and if you know that the module will fail without check mode, it should also fail with check mode. (There are always situations where determining whether it fails is complicated, and it's not done because of that, but in this case it's very easy :) )

pratikgadiya12 replied:
noted, fixing this

pratikgadiya12 replied:
fixed

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +373,65 @@ def main():
+    elif state == "import":
+        changed = True
+        if module.check_mode:
+            module.exit_json(changed=True, db=db_name, db_list=db)
```
```suggestion
            module.exit_json(changed=bool(non_existence_list), db=db_name, db_list=db)
```

pratikgadiya12 replied:
Let's consider a situation where all the databases to be imported already exists. In this case, the changed state for check mode would be `False` since `changed=bool(non_existence_list)`. This contradicts to the previous `mysql_db` code for import where it always returns `True` irrespective of whether a database exists or not.

Previous code for `state=import`:
- db exists [code](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/mysql/mysql_db.py#L372)
- db not exist [code](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/mysql/mysql_db.py#L400)

felixfontein replied:
You're right, it always returned `True` before. There's a difference to before, though: before, `db_create` *always* returned `True`. Now, it only returns `True` if creating all DBs succeeded. So the `if changed:` can be removed in line 426, and the `if` block can always be executed. As long as `db_create` didn't result in an exception (in which case the module dies), all DBs should be there anyway.

pratikgadiya12 replied:
Agree that `if changed:` could be removed and `if` block can always be executed however, I still think that `changed` should be equals to `True` instead of `changed=bool(non_existence_list)` since import functionality also overrides imported database if that already exists.
This will again lead to system change, right?

pratikgadiya12 replied:
Discussed the same with @felixfontein and we agree that `changed` state would be always `True`

pratikgadiya12 committed:
Code refactoring and review fixes

- Added dump all test case

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -318,15 +345,14 @@ def main():
-    if len(db) > 1 and state != 'dump':
-        module.fail_json(msg="Multiple databases is only supported with state=dump")
+    if len(db) > 1 and state == 'import':
+        module.fail_json(msg="Multiple databases is not supported with state=import")
```
```suggestion
        module.fail_json(msg="Multiple databases are not supported with state=import")
```

pratikgadiya12 replied:
fixed

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +372,65 @@ def main():
+
+    existence_list = []
+    non_existence_list = []
+    for each_database in db:
```
Maybe skip this `for` if `all_databases` is `True`?

pratikgadiya12 replied:
fixed

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +372,65 @@ def main():
+        if non_existence_list:
             try:
-                changed = db_delete(cursor, db)
+                changed = db_create(cursor, non_existence_list, encoding, collation)
```
You want `changed` to be `True` here if *at least* one DB was created, not if *all* DBs in `non_existence_list` were created (and there has been at least one).

pratikgadiya12 replied:
fixed

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +372,65 @@ def main():
+        if non_existence_list and not all_databases:
+            module.fail_json(msg="Cannot dump database(s) %r - not found" % (', '.join(non_existence_list)))
+        if module.check_mode:
+            module.exit_json(changed=not bool(non_existence_list), db=db_name, db_list=db)
```
```suggestion
            module.exit_json(changed=True, db=db_name, db_list=db)
```
So behavior is the same as without `check_mode`.

pratikgadiya12 replied:
fixed

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +372,65 @@ def main():
+            except Exception as e:
+                module.fail_json(msg="error creating database: %s" % to_native(e),
+                                 exception=traceback.format_exc())
+        if changed:
```
Remove this line and de-dent `if:` block. (And remove the `changed` variable in the `state == 'import'` block completely.)

pratikgadiya12 replied:
fixed, missed this in last commit

felixfontein commented on changelogs/fragments/58370-mysqldb-allow-multiple_databases_create_delete_operation.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- mysql_db now supports creation and deletion of multiple databases (https://github.com/ansible/ansible/issues/58370)
```
I would add another line that it not behaves better w.r.t. `changed` result in `check_mode`.

pratikgadiya12 replied:
done

pratikgadiya12 committed:
Fixed review comments

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -396,6 +397,8 @@ def main():
         if non_existence_list:
             try:
                 changed = db_create(cursor, non_existence_list, encoding, collation)
+                if changed >= 1:
```
`changed` is a boolean, not an integer.

pratikgadiya12 replied:
Good catch, fixed

felixfontein commented on changelogs/fragments/58370-mysqldb-allow-multiple_databases_create_delete_operation.yml:
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+- mysql_db now supports creation and deletion of multiple databases (https://github.com/ansible/ansible/issues/58370)
+- Does not behave better w.r.t `changed` results in `check_mode`
```
```suggestion
- mysql_db - now behaves better w.r.t ``changed`` results in ``check_mode``
```

pratikgadiya12 replied:
Fixed

pratikgadiya12 committed:
Minor review comment fixes

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +378,68 @@ def main():
+                module.fail_json(msg="error creating database: %s" % to_native(e),
+                                 exception=traceback.format_exc())
+            # If atleast one database was created by db_create then changed should be True
+            if not changed and any_db_exists(non_existence_list):
```
I don't think this works. In case `db_create` returns `False`, it can be that some of the DBs were created, or none (because they were already created between the existence check and the CREATE DATABASE command). In case `db_create` didn't fail and `non_existence_list` is not empty, `any_db_exists(non_existence_list)` will *always* return `True`. But `changed` might still have to be `False`.

pratikgadiya12 replied:
fixed with the suggestion provided in comment https://github.com/ansible/ansible/pull/58602#discussion_r302730546

felixfontein commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -248,15 +270,26 @@ def db_import(module, host, user, password, db_name, target, all_databases, port
+            query.append("COLLATE %(collate)s")
+        query = ' '.join(query)
+        res += cursor.execute(query, query_params)
+    return res == len(db)
```
You should change this one to
```suggestion
    return res > 0
```
to avoid the problem. The function `db_create` is called only in two places; in one place, the return value is ignored (import), and in the other (present), you need to know whether *at least one* DB was created, and not whether *all* DBs have been created. So if you return `res > 0` here, `changed` is determined correctly for `present`.

pratikgadiya12 replied:
fixed

pratikgadiya12 committed:
Altered db_create return value

felixfontein reviewed (approved):
I think the code looks good now. @Andersson007 feel free to continue your review :)

Andersson007 commented:
@felixfontein and @pratikgadiya12  , thanks, will start Monday morning :)

felixfontein commented:
@Andersson007 have a great weekend until then! :)

Andersson007 commented:
@felixfontein, have a nice weekend too !

Andersson007 commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -293,7 +320,8 @@ def main():
     if not db:
-        module.fail_json(msg="Please provide at least one database name")
+        module.exit_json(changed=False, db=db)
+    db = [each_db.strip() for each_db in db]
```
good

Andersson007 commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +372,65 @@ def main():
+            changed = db_delete(cursor, existence_list)
+        except Exception as e:
+            module.fail_json(msg="error deleting database: %s" % to_native(e))
+        module.exit_json(changed=changed, db=db_name, db_list=db)
```
it's repeated at the and of each condition. Maybe we could use it one time after
```suggestion
        module.exit_json(changed=changed, db=db_name, db_list=db)
```

pratikgadiya12 replied:
Actually `exit_json` is not same for all 4 state's. Its same for state=absent/present or state=dump/import.

Andersson007 commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -346,81 +372,65 @@ def main():
+
+    if state == "absent":
+        if module.check_mode:
+            module.exit_json(changed=bool(existence_list), db=db_name, db_list=db)
```
looks like db and db_list are the same, doesn't it? If it does, I think one of them should be removed (probably db_list)

pratikgadiya12 replied:
Their values are same, however, the `type` is different.

I kept `db` and `db_list` since I didn't want to break existing playbooks of users which explicitly checks the content of `result.db` as [here](https://app.shippable.com/github/ansible/ansible/runs/132137/55/console) eventually breaking their playbook.
```
01:19 fatal: [testhost]: FAILED! => {
01:19     "assertion": "result.db =='data'",
01:19     "changed": false,
01:19     "evaluated_to": false,
01:19     "msg": "Assertion failed"
01:19 }
```

I need your opinion on this to know if I should be keeping both the variables or should I just be returning list of databases
```
-        module.exit_json(changed=changed, db=db_name, db_list=db)
+        module.exit_json(changed=changed, db=db)
```

felixfontein replied:
I think keeping both is probably best. Changing types of an existing return value, or removing one, will definitely break some people's playbooks.

You could deprecate the old return value and announce that in the porting guide; then the old return value can be dropped in 2.13. This is not optimal because there is no mechanism which tells people that the result they are using is deprecated, but at least people who read the porting guide and/or changelog will have four versions to adapt their playbooks.

Andersson007 replied:
@felixfontein , I thought that it was added by this PR, if it was there before, sure, it must be kept !
@pratikgadiya12 , dont remove it, please! :)

felixfontein replied:
@Andersson007 I interpreted @pratikgadiya12's statement this way. Now I also checked, it was there before, indeed. What's completely missing (and has been in the past) is documentation for the return values! @pratikgadiya12 can you add this? I guess `db` has been there all the time, and `db_list` will be `version_added: 2.9`.

felixfontein replied:
Ah, I just saw you already mentioned the return value docs :)

pratikgadiya12 replied:
Will keep both the return values and add the return value docs

Andersson007 commented on test/integration/targets/mysql_db/tasks/multi_db_create_delete.yml:
```diff
@@ -0,0 +1,699 @@
+  command: mysql "-e show databases like 'database%';"
+  register: mysql_result
+
+- name: assert that databases does exist after creation
```
```suggestion
- name: assert that databases exist after creation
```

pratikgadiya12 replied:
fixed

Andersson007 commented on test/integration/targets/mysql_db/tasks/multi_db_create_delete.yml:
```diff
@@ -0,0 +1,699 @@
+  command: mysql "-e show databases like 'database%';"
+  register: mysql_result
+
+- name: assert that databases does exist (since performed recreation of existing databases via check mode)
```
```suggestion
- name: assert that databases exist (since performed recreation of existing databases via check mode)
```

Andersson007 replied:
etc

pratikgadiya12 replied:
Altered "does exist" to just "exist" wherever found

Andersson007 commented on test/integration/targets/mysql_db/tasks/multi_db_create_delete.yml:
```diff
@@ -0,0 +1,699 @@
+      - "'{{ db5_name }}' not in mysql_result.stdout"
+
+- name: state dump - file name should exist
+  file: name={{ dump1_file }} state=file
```
```suggestion
  file: name={{ dump1_file }} state=file
```
we could also grep the dump file by shell module to be sure that there are something like CREATE DATABASE blabla

pratikgadiya12 replied:
fixed

Andersson007 commented on test/integration/targets/mysql_db/tasks/multi_db_create_delete.yml:
```diff
@@ -0,0 +1,699 @@
+      - "'{{ db4_name }}' not in mysql_result.stdout"
+      - "'{{ db5_name }}' not in mysql_result.stdout"
+
+# ==============================IMPORT OPERATIONS ==========================
```
```suggestion
# ==============================IMPORT OPERATIONS ==========================
```
I'd like to see it in another test file because it isn't related to the case

pratikgadiya12 replied:
Moved it to state_dump_import.yml file

Andersson007 commented:
thank you, guys, for the great work!
two points about tests:
1. IMO better to move tests that related to import to another file (suitable existent or new)
2. there are duplicated tests, @pratikgadiya12 , if you check, for example, deletion, you don't need to do it again each time when you need to delete databases for the next steps, etc. So, better to reduce amount of test code by deleting assertions that were before (you could remark in test comment that you delete databases because it's needed for the next step, etc.)

Andersson007 commented on test/integration/targets/mysql_db/tasks/multi_db_create_delete.yml:
```diff
@@ -0,0 +1,699 @@
+  command: mysql "-e show databases like 'database%';"
+  register: mysql_result
+
+- name: assert that databases does exist
```
```suggestion
- name: assert that databases exist
```
I'm not a native speaker but it looks to me like a typo (I saw it several times)

pratikgadiya12 replied:
altered

pratikgadiya12 committed:
Removed db_list and altered "does exist" to just "exist"

pratikgadiya12 committed:
Kept db and db_list in module.exit_json

pratikgadiya12 committed:
Refactored tests

- Added removal of dump2 file

pratikgadiya12 committed:
Moved import tests to state_dump_import file

pratikgadiya12 committed:
Removed import tests from multi_db_create_delete

pratikgadiya12 committed:
Updated porting guide, added RETURN block

pratikgadiya12 committed:
Minor identation fix

Andersson007 commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -115,6 +122,32 @@
+
+RETURN = r'''
+db:
+  description: Database names in string format delimted by white space.
```
```suggestion
  description: Database names in string format delimited by white space.
```

pratikgadiya12 replied:
fixed typo

pratikgadiya12 committed:
Added validation to check if databases are dumped

pratikgadiya12 commented:
> thank you, guys, for the great work!
> two points about tests:
> 
> 1. IMO better to move tests that related to import to another file (suitable existent or new)
> 2. there are duplicated tests, @pratikgadiya12 , if you check, for example, deletion, you don't need to do it again each time when you need to delete databases for the next steps, etc. So, better to reduce amount of test code by deleting assertions that were before (you could remark in test comment that you delete databases because it's needed for the next step, etc.)
> 3. The module return db/db_list so, IMO better to add RETURN = r'''...''' block (see almost any postgresql modules) and check it into the tests, e.g.:
> 
> ```
> - assert:
>     that:
>     - result is changed
>     - result.db == [..., ..., ...]
> ```

Thanks for the review @felixfontein and @Andersson007!
I have fixed all the above review comments

Andersson007 reviewed (approved):
shipit

Andersson007 commented:
@pratikgadiya12 thanks for the contribution, great work, thanks!
@felixfontein , thanks for reviewing!

pratikgadiya12 commented:
Thanks a lot @felixfontein and @Andersson007 for great review!

We could merge this @felixfontein since this PR has 2 shipit's from you guys :)

Andersson007 commented:
!needs_revision

bmalynovytch commented:
bot_status

bmalynovytch commented on lib/ansible/modules/database/mysql/mysql_db.py:
```diff
@@ -136,13 +169,16 @@
     res = 0
     for each_db in db:
-        res += cursor.execute("SHOW DATABASES LIKE %s", (each_db.strip().replace("_", r"\_"),))
+        res += cursor.execute("SHOW DATABASES LIKE %s", (each_db.replace("_", r"\_"),))
```
Good point for handling the `_` char, but have you tested the `-` char ?
I know some devs use it 🙄 in the db name, as long as they quote the db name.

pratikgadiya12 replied:
@bmalynovytch: Yes, it does work for `-` char as well.

**Ansible-playbook**:
```
$ cat only_create.yml 
---
- name: Sample
  gather_facts: no
  hosts: localhost
  tasks:
  - name: Create multiple databases
    mysql_db:
      name:
        - "abcd-pratik"
      state: present
      login_host: 172.17.0.2
      login_user: monty
      login_password: pw
    register: result
  - debug:
      msg: "{{ result }}"
```

**Ansible Run**:
```
$ ansible-playbook only_create.yml 
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [Sample] *****************************************************************************************************************************************************************************************************

TASK [Create multiple databases] **********************************************************************************************************************************************************************************
changed: [localhost]

TASK [debug] ******************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": {
        "changed": true, 
        "db": "abcd-pratik", 
        "db_list": [
            "abcd-pratik"
        ], 
        "failed": false
    }
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

Similarly tested it for `delete` operation as well.

bmalynovytch replied:
Marvelous ! 👏

bmalynovytch commented:
shipit

felixfontein commented:
@pratikgadiya12 thanks a lot for implementing this!
@bmalynovytch @Andersson007 thanks a lot for reviewing!
